README.md
# ANPR
## Licence plate recognition for Homey ##
Use snapshots from a camera connected to Homey, and let Platerecognizer do AI based licence plate recognition. For every plate the app detects, a flow will be triggered. The app will also provide vehicle type and region.

<img src="https://aws1.discourse-cdn.com/business4/uploads/athom/original/3X/6/f/6f0957c6d850fb69e8c70c1838c33abe8027ca8a.jpeg" alt="ANPR" width="250">

## Use cases ##
* Presence detection
* Security
* Open garage door
* Fun

## Privacy ##
The Homey app uses a cloud based licence plate recognition service called Platerecognizer. Using the Platerecognizer service has privacy implications. Use at own risk. Read and understand your local privacy legislation, and the terms of use from Platerecognizer.

## Performance ##
The free Platerecognizer service allows you to upload 1 or 2 images per second. The plate detection usually is ready within 1 to 3 seconds from uploading an image (via an action flowcard).

## Test image ##
To make testing your flow easier, a test image is available as global image token. It has license plates of two cars and a motor cycle.

<img src="https://aws1.discourse-cdn.com/business4/uploads/athom/original/3X/3/a/3ad14d18b336841922871f43b577d74b1bdd893c.jpeg" alt="ANPR" width="250">



changelog.txt
v2.0.0  01.11.2021 Migration to SDK3. Fix HTTPS timeout. Improved logger. Removed % used tokens 2 through 4.
v1.1.4  10.06.2020 Removed multiple API key support.
v1.1.3  12.05.2020 Memory leak fix. Faster statistics update.
v1.1.1  10.05.2020 Added multiple API key support. Added key validity check on save. Added API statistics. Added manual region selection. Bug fixes.
v1.0.1  05.05.2020 Added toggle API Secret view in app settings. Fix API throttle issue. Fix threshold setting.
v1.0.0  04.05.2020 First release.

app.js
/*
Copyright 2020 -2021, Robin de Gruijter (gruijter@hotmail.com)

This file is part of com.gruijter.anpr.

com.gruijter.anpr is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

com.gruijter.anpr is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with com.gruijter.anpr.  If not, see <http://www.gnu.org/licenses/>.
*/

'use strict';

const Homey = require('homey');
// const util = require('util');
// const fs = require('fs');

const Logger = require('./captureLogs.js');
const ANPR = require('./platerecognizer.js');

// const setTimeoutPromise = util.promisify(setTimeout);

// convert binary data to base64 encoded string
const base64Encode = (img) => Buffer.from(img).toString('base64');
// const base64Decode = (base64) => Buffer.from(base64, 'base64');

class ANPRApp extends Homey.App {

	async onInit() {
		try {
			if (!this.logger) this.logger = new Logger({ name: 'log', length: 200, homey: this.homey });

			// register some listeners
			process.on('unhandledRejection', (error) => {
				this.error('unhandledRejection! ', error);
			});
			process.on('uncaughtException', (error) => {
				this.error('uncaughtException! ', error);
			});
			this.homey
				.on('unload', () => {
					this.log('app unload called');
					// save logs to persistant storage
					this.logger.saveLogs();
				})
				.on('memwarn', () => {
					this.error('memwarn!');
				});

			// first remove settings events listener
			if (this.listeners && this.listeners.set) {
				this.homey.settings.removeListener('set', this.listeners.set);
			}

			// listen to change settings events
			this.listeners.set = (key) => {
				if (key === 'settingsMatch') {
					this.log('Plate search settings changed:', JSON.stringify(this.homey.settings.get('settingsMatch')));
				}
				if (key === 'settingsKey') {
					this.log('API Key settings changed, reloading app now');
				}
				this.logger.saveLogs();
				this.log('Re-initing app now');
				this.onInit();
			};
			this.homey.settings
				.on('set', this.listeners.set);

			// init values from app settings
			const matchSettings = this.homey.settings.get('settingsMatch') || {}; // { threshold, regions };
			this.matchThreshold = (matchSettings.threshold || 85) / 100;
			this.regions = matchSettings.regions;	// [ 'nl', 'de' ]

			const keySettings = this.homey.settings.get('settingsKey') || {}; // { apiKeys: [ apiKey1 ... apiKey4 ] }
			let { apiKeys } = keySettings;
			apiKeys = apiKeys ? apiKeys.filter((key) => key.length > 0) : [];

			const options = { apiKeys };
			this.ANPR = new ANPR(options);

			if (!apiKeys || !apiKeys[0]) {
				this.log('No API key entered in app settings');
				return;
			}

			// register flows and tokens, and start polling stuff
			await this.registerFlowListeners();
			await this.registerFlowTokens();
			// this.startPolling();

			// get stats
			this.updateStats();

			this.log('ANPR is running...');

			// do garbage collection every 10 minutes
			// this.intervalIdGc = setInterval(() => {
			// 	global.gc();
			// }, 1000 * 60 * 10);

		} catch (error) {
			this.error(error);
		}

	}

	//  stuff for frontend API
	deleteLogs() {
		return this.logger.deleteLogs();
	}

	getLogs() {
		return this.logger.logArray;
	}

	checkKey(body) {
		this.log('Checking new API key from frontend');
		return Promise.resolve(this.ANPR.getStatistics(body.key));
	}

	// update usage statistics tokens
	async updateStats() {
		try {
			if (this.busyStat) return;
			this.busyStat = true;
			const stats = await this.ANPR.getAllStatistics();
			stats.forEach((stat, index) => {
				const token = `key${index + 1}UsageToken`;
				if (stat && stat.total_calls && this.tokens && this.tokens[token]) {
					const keyUsage = Math.round((stat.usage.calls / stat.total_calls) * 1000) / 10;
					this.tokens[token].setValue(keyUsage);
				}
			});
			this.busyStat = false;
		} catch (error) {
			this.error(error);
		}
	}

	// returns a formatted array of recognised plates, and triggers flowCards
	async detectPlates(imgBase64, origin) {
		try {
			const options = {
				upload:	imgBase64,	// (required) The file to be uploaded.
				regions: this.regions, // Match the license plate pattern of a specific region or regions.
				camera_id:	origin || 'homey',	// Unique camera identifier.
				timestamp:	new Date().toISOString(),	// ISO 8601 timestamp. For example, 2019-08-19T13:11:25. The timestamp has to be in UTC.
			};
			const { results, camera_id: cameraID } = await this.ANPR.detectPlates(options);
			// console.dir(results, { depth: null });
			const searchResult = results
				.filter((plate) => plate.score > this.matchThreshold)
				.map((plate) => {
					const tokens = {
						origin: cameraID, // || origin || 'undefined',
						plate: plate.plate,
						confidence: Math.round(plate.score * 1000) / 10,
						quality: Math.round(plate.dscore * 1000) / 10,
						region: plate.region.code,
						region_conf: Math.round(plate.region.score * 1000) / 10,
						vehicle: plate.vehicle.type,
						vehicle_conf: Math.round(plate.vehicle.score * 1000) / 10,
					};
					this.log(tokens);
					const plateDetectedTrigger = this.homey.flow.getTriggerCard('plate_detected');
					plateDetectedTrigger.trigger(tokens);
					return tokens;
				});
			await Promise.all(searchResult);
			if (searchResult.length === 0) this.log(`no plates found in image ${origin}`);
			this.updateStats();
		} catch (error) {
			this.error(error.message);
		}
	}

	async analyzeImage(args) {
		try {
			// get the contents of the image
			const image = args.droptoken;
			if (!image || !image.getStream) throw Error('no valid image stream');
			// if (typeof image === 'undefined' || image == null) return false;
			const imageStream = await image.getStream();
			// load image in memory
			imageStream.once('error', (err) => {
				this.error(err);
			});
			const chunks = [];
			// File is done being read
			imageStream.once('end', () => {
				const imgBuffer = Buffer.concat(chunks);
				this.detectPlates(base64Encode(imgBuffer), args.origin);
			});
			imageStream.on('data', (chunk) => {
				chunks.push(chunk); // push data chunk to array
			});
		} catch (error) {
			this.error(error);
		}
	}

	// register flow cards
	async registerFlowListeners() {
		try {
			// action cards
			const analyzeImageAction = this.homey.flow.getActionCard('search_plates');
			analyzeImageAction.registerRunListener((args) => this.analyzeImage(args));
			return Promise.resolve(this.flows);
		} catch (error) {
			return Promise.resolve(error);
		}
	}

	// register global flow tokens
	async registerFlowTokens() {
		try {
			// unregister tokens first
			if (!this.tokens) this.tokens = {};
			const ready = Object.keys(this.tokens).map((token) => Promise.resolve(this.homey.ManagerFlow.unregisterToken(this.tokens[token])));
			await Promise.all(ready);

			// register the test image
			const testImage = await this.homey.images.createImage();
			testImage.setPath('/assets/images/test.jpg');

			// create a droptoken for test image
			this.tokens.testImageToken = await this.homey.flow.createToken('test_image_token', {
				type: 'image',
				title: 'Test Image',
			});
			await this.tokens.testImageToken.setValue(testImage);

			// register API call usage token
			this.tokens.key1UsageToken = await this.homey.flow.createToken('key1_usage_token', {
				type: 'number',
				title: 'Key usage %',
			});

			return Promise.resolve(this.tokens);
		} catch (error) {
			return Promise.resolve(error);
		}

	}

	// register polling stuff
	// async startPolling() {
	// 	try {
	// 		// clear intervals first
	// 		if (!this.interval) this.interval = {};
	// 		const ready = Object.keys(this.interval).map((interval) => Promise.resolve(clearInterval(this.interval[interval])));
	// 		await Promise.all(ready);

	// 		// start polling statistics every minute
	// 		this.interval.stats = setInterval(() => {
	// 			this.updateStats();
	// 		}, 1000 * 60 * 0.1);

	// 	} catch (error) {
	// 		this.error(error);
	// 	}

	// }

}

module.exports = ANPRApp;

/*

{
  origin: 'undefined',
  plate: 'dj1313',
  confidence: 90.3,
  quality: 80.9,
  region: 'lu',
  region_conf: 86.4,
  vehicle: 'Car',
  vehicle_conf: 80.2
}

[
  { total_calls: 2500, usage: { year: 2020, month: 5, calls: 832 } },
  { total_calls: 2500, usage: { year: 2020, month: 5, calls: 13 } }
]

*/


captureLogs.js
/*
Copyright 2020, Robin de Gruijter (gruijter@hotmail.com)

This file is part of com.gruijter.anpr.

com.gruijter.anpr is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

com.gruijter.anpr is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with com.gruijter.anpr.  If not, see <http://www.gnu.org/licenses/>.
*/

'use strict';

const Homey = require('homey');
const StdOutFixture = require('fixture-stdout');
const fs = require('fs');

class captureLogs {
	// Log object to keep logs in memory and in persistent storage
	// captures and reroutes Homey's this.log (stdout) and this.err (stderr)

	constructor(logName, logLength) {
		this.logName = logName || 'log';
		this.logLength = logLength || 50;
		this.logFile = `/userdata/${this.logName}.json`;
		this.logArray = [];
		this.getLogs();
		this.captureStdOut();
		this.captureStdErr();
	}

	getLogs() {
		try {
			const log = fs.readFileSync(this.logFile, 'utf8');
			this.logArray = JSON.parse(log);
			Homey.app.log('logfile retrieved');
			return this.logArray;
		} catch (error) {
			if (error.message.includes('ENOENT')) return [];
			Homey.app.error('error parsing logfile: ', error.message);
			return [];
		}
	}

	saveLogs() {
		try {
			fs.writeFileSync(this.logFile, JSON.stringify(this.logArray));
			Homey.app.log('logfile saved');
			return true;
		} catch (error) {
			Homey.app.error('error writing logfile: ', error.message);
			return false;
		}
	}

	deleteLogs() {
		try {
			fs.unlinkSync(this.logFile);
			this.logArray = [];
			Homey.app.log('logfile deleted');
			return true;
		} catch (error) {
			if (error.message.includes('ENOENT')) return false;
			Homey.app.error('error deleting logfile: ', error.message);
			return false;
		}
	}

	captureStdOut() {
		// Capture all writes to stdout (e.g. this.log)
		this.captureStdout = new StdOutFixture({ stream: process.stdout });
		this.captureStdout.capture((string) => {
			if (this.logArray.length >= this.logLength) {
				this.logArray.shift();
			}
			this.logArray.push(string);
			// return false;	// prevent the write to the original stream
		});
		Homey.app.log('capturing stdout');
		// captureStdout.release();
	}

	captureStdErr() {
		// Capture all writes to stderr (e.g. this.error)
		this.captureStderr = new StdOutFixture({ stream: process.stderr });
		this.captureStderr.capture((string) => {
			if (this.logArray.length >= this.logLength) {
				this.logArray.shift();
			}
			this.logArray.push(string);
			// return false;	// prevent the write to the original stream
		});
		Homey.app.log('capturing stderr');
		// captureStderr.release();
	}

	releaseStdOut() {
		this.captureStdout.release();
	}

	releaseStdErr() {
		this.captureStderr.release();
	}

}

module.exports = captureLogs;


settings/settings.js
/* eslint-disable no-unused-vars */
/* eslint-disable no-undef */

// tab 2 stuff here
function displayLogs(lines) {
	$('#loglines').html(lines);
}

function updateLogs() {
	try {
		displayLogs('');
		const showLogs = $('#show_logs').prop('checked');
		const showErrors = $('#show_errors').prop('checked');
		const showNoDetects = $('#show_no_detections').prop('checked');
		Homey.api('GET', 'getlogs/', null, (err, result) => {
			if (!err) {
				let lines = '';
				result
					.reverse()
					.forEach((line) => {
						if (!showLogs) {
							if (line.includes('[log]')) return;
						}
						if (!showErrors) {
							if (line.includes('[err]')) return;
						}
						if (!showNoDetects) {
							if (line.includes(' found in image')) return;
						}
						const logLine = line.replace(' [ANPRApp]', '');
						lines += `${logLine}<br />`;

					});
				displayLogs(lines);
			} else {
				displayLogs(err);
			}
		});
	} catch (e) {
		displayLogs(e);
	}
}

function deleteLogs() {
	Homey.confirm(Homey.__('settings.tab2.deleteWarning'), 'warning', (error, result) => {
		if (result) {
			Homey.api('GET', 'deletelogs/', null, (err) => {
				if (err) {
					Homey.alert(err.message, 'error'); // [, String icon], Function callback )
				} else {
					Homey.alert(Homey.__('settings.tab2.deleted'), 'info');
					updateLogs();
				}
			});
		}
	});
}

// tab 3 stuff here
function checkKeys(set) {
	try {
		const ready = set.map((key) => new Promise((resolve, reject) => {
			Homey.api('POST', 'checkkey/', { key }, (err, result) => {
				if (err) return reject(err);
				return resolve(result);
			});
		}));
		return Promise.all(ready);
	} catch (error) {
		return Promise.reject(error);
	}
}

async function showInfo3() {
	try {
		const key1 = document.getElementById('apiKey1');
		// const key2 = document.getElementById('apiKey2');
		// const key3 = document.getElementById('apiKey3');
		// const key4 = document.getElementById('apiKey4');
		key1.type = 'password';
		// key2.type = 'password';
		// key3.type = 'password';
		// key4.type = 'password';
		Homey.get('settingsKey', (err, set) => {
			if (err || !set || !set.apiKeys) return;
			$('#apiKey1').val(set.apiKeys[0]);
			// $('#apiKey2').val(set.apiKeys[1]);
			// $('#apiKey3').val(set.apiKeys[2]);
			// $('#apiKey4').val(set.apiKeys[3]);
		});
		Homey.get('settingsMatch', (err, set) => {
			if (err || !set || !set.regions) return;
			$('#threshold').val(set.threshold);
			$('#regions').val(set.regions.join());
		});
	} catch (error) {
		Homey.alert(error.message, 'error');
	}
}

function togglePasswordView() {
	const key1 = document.getElementById('apiKey1');
	// const key2 = document.getElementById('apiKey2');
	// const key3 = document.getElementById('apiKey3');
	// const key4 = document.getElementById('apiKey4');
	if (key1.type === 'password') {
		key1.type = 'text';
		// key2.type = 'text';
		// key3.type = 'text';
		// key4.type = 'text';
	} else {
		key1.type = 'password';
		// key2.type = 'password';
		// key3.type = 'password';
		// key4.type = 'password';
	}
}

async function saveSettingsKey() {
	try {
		const apiKeys = [$('#apiKey1').val()] // , $('#apiKey2').val(), $('#apiKey3').val(), $('#apiKey4').val()]
			.filter((key) => key.length > 0);
		if (!apiKeys[0]) return Homey.alert('API Key must be entered!', 'error');
		await checkKeys(apiKeys);
		await Homey.set('settingsKey', { apiKeys });
		return Homey.alert('API Key is saved!', 'info');
	} catch (error) {
		return Homey.alert(error.message, 'error');
	}
}

async function saveSettingsMatch() {
	try {
		const threshold = $('#threshold').val();
		if (threshold < 0 || threshold > 100) throw Error('Threshold must be between 0 and 100');
		const regions = $('#regions').val().replace(' ', '').split(',');
		await Homey.set('settingsMatch', { threshold, regions });
		return Homey.alert('Settings are saved!', 'info');
	} catch (error) {
		return Homey.alert(error.message, 'error');
	}
}

// generic stuff here
function showTab(tab) {
	if (tab === 2) updateLogs();
	if (tab === 3) showInfo3();

	$('.tab').removeClass('tab-active');
	$('.tab').addClass('tab-inactive');
	$(`#tabb${tab}`).removeClass('tab-inactive');
	$(`#tabb${tab}`).addClass('active');
	$('.panel').hide();
	$(`#tab${tab}`).show();
}

function onHomeyReady(homeyReady) {
	Homey = homeyReady;
	showInfo3();
	showTab(1);
	Homey.ready();
}


platerecognizer.js
/*
Copyright 2020 - 2021, Robin de Gruijter (gruijter@hotmail.com)

This file is part of com.gruijter.anpr.

com.gruijter.anpr is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

com.gruijter.anpr is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with com.gruijter.anpr.  If not, see <http://www.gnu.org/licenses/>.
*/

'use strict';

const https = require('https');
const qs = require('querystring');
const Queue = require('smart-request-balancer');
// const util = require('util');

const APIHost = 'api.platerecognizer.com';
const detectLPlateEP = '/v1/plate-reader/';
const statisticsEP = '/v1/statistics/';

const parse = (data) => {
	try {
		const parsed = JSON.parse(data);
		return parsed;
	} catch (error) {
		return {};
	}
};

class ANPR {

	constructor(opts) {
		const options = opts || {};
		this.host = options.host || APIHost;
		this.port = options.port || 443;
		this.timeout = options.timeout || 10000;
		this.apiKeys = options.apiKeys || [];
		this.apiKeyIndex = 0;
		this.initQueue();
	}

	// queue stuff
	initQueue() {
		const config = {
			rules: {				// Describing our rules by rule name
				common: {			// Common rule. Will be used if you won't provide rule argument
					rate: 2,		// Allow to send 10 messages
					limit: 1,		// per 1 second
					priority: 1,	// Rule priority. The lower priority is, the higher chance that this rule will execute faster
				},
			},
			overall: {				// Overall queue rates and limits
				rate: 10,
				limit: 1,
			},
			retryTime: 2,		// Default retry time (seconds). Can be configured in retry fn
			ignoreOverallOverheat: true,	// Should we ignore overheat of queue itself
		};
		this.queue = new Queue(config);
	}

	queueMessage(path, msg, apiKey) {
		const key = Date.now(); // 'homey';
		const requestHandler = (retry) => this._makeRequest(path, msg, apiKey)
			.then((response) => response)
			.catch((error) => {
				if (error.message && (error.message.includes('throttled') || error.message.includes('429'))) {	// or 500?
					return retry();
				}
				throw error;
			});
		return this.queue.request(requestHandler, key);
	}

	// returns an array of detected plates with attributes
	async detectPlates(opts) {
		try {
			const randomIndex = Math.floor(Math.random() * this.apiKeys.length);
			const apiKey = this.apiKeys[randomIndex];

			const postData = {
				// upload:	(required) The file to be uploaded. The parameter can either be the file bytes (using Content-Type multipart/form-data) OR a base64 encoded image.
				// regions: ['nl', 'be', 'de'], // Match the license plate pattern of a specific region or regions. This parameter can be used multiple times to specify more than one region.
				// camera_id:	Unique camera identifier.
				// timestamp:	new Date().toISOString(),	// ISO 8601 timestamp. For example, 2019-08-19T13:11:25. The timestamp has to be in UTC.
			};
			Object.assign(postData, opts);
			const result = await this.queueMessage(detectLPlateEP, postData, apiKey);
			// const result = await this._makeRequest(detectLPlateEP, postData, apiKey);
			return Promise.resolve(result);
		} catch (error) {
			return Promise.reject(error);
		}
	}

	// returns the statistics of the used API key
	async getStatistics(apiKey) {
		try {
			const stat = await this.queueMessage(statisticsEP, '', apiKey);
			return Promise.resolve(stat);
		} catch (error) {
			return Promise.reject(error);
		}
	}

	getAllStatistics() {
		try {
			const stats = this.apiKeys.map((apiKey) => this.getStatistics(apiKey).catch(() => null));
			return Promise.all(stats);
		} catch (error) {
			return Promise.reject(error);
		}
	}

	// HTTPS request

	async _makeRequest(path, msg, apiKey) {
		try {
			const message = qs.stringify(Object.assign(msg));
			const headers = {
				Authorization: `Token ${apiKey}`,
				// 'cache-control': 'no-cache',
				// 'content-type': 'multipart/form-data',	// for stream
				'content-type': 'application/x-www-form-urlencoded',	// for base64 encoded file
				'content-length': Buffer.byteLength(message),
				connection: 'Keep-Alive',
			};
			const options = {
				hostname: this.host,
				port: this.port,
				path,
				headers,
				method: 'POST',
			};
			if (path === statisticsEP) options.method = 'GET';
			const result = await this._makeHttpsRequest(options, message);
			const body = parse(result.body);
			// console.log(util.inspect(body, false, 10, true));
			if (!Object.keys(body).length) throw Error(result.statusCode, result.body);
			// if (body.detail) throw Error(body.detail);
			// if (body.non_field_errors) throw Error(body.non_field_errors[0]);
			// if (body.upload) throw Error(body.upload[0]);
			if (result.statusCode !== 200 && result.statusCode !== 201) throw Error(JSON.stringify(body));
			// all is good
			return Promise.resolve(body);
		} catch (error) {
			return Promise.reject(error);
		}
	}

	_makeHttpsRequest(options, postData, timeout) {
		return new Promise((resolve, reject) => {
			const opts = options;
			opts.timeout = timeout || this.timeout;
			const req = https.request(opts, (res) => {
				let resBody = '';
				res.on('data', (chunk) => {
					resBody += chunk;
				});
				res.once('end', () => {
					this.lastResponse = resBody;
					if (!res.complete) {
						return reject(Error('The connection was terminated while the message was still being sent'));
					}
					res.body = resBody;
					return resolve(res);
				});
			});
			req.on('error', (e) => {
				req.destroy();
				this.lastResponse = e;
				return reject(e);
			});
			req.on('timeout', () => {
				req.destroy();
			});
			req.end(postData);
		});
	}

}

module.exports = ANPR;

/*

{
  "usage": {
    "month": 1,
    "calls": 128,
    "year": 2019
  },
  "total_calls": 2500
}

{
  processing_time: 67.586,
  results: [
    {
      box: { xmin: 124, ymin: 459, xmax: 292, ymax: 509 },
      plate: 'dj1313',
      region: { score: 0.864, code: 'lu' },
      vehicle: {
        score: 0.802,
        box: { xmin: 36, ymin: 174, xmax: 833, ymax: 597 },
        type: 'Car'
      },
      score: 0.903,
      candidates: [
        { score: 0.903, plate: 'dj1313' },
        { score: 0.761, plate: 'dj13i3' },
        { score: 0.761, plate: 'dji313' },
        { score: 0.619, plate: 'dji3i3' },
        [length]: 4
      ],
      dscore: 0.809
    },
    [length]: 1
  ],
  filename: '2009_sb90q_f4d06171-f321-4f23-bd96-da7a8cd577d4.jpg',
  version: 1,
  camera_id: 'homey',
  timestamp: '2020-05-03T20:09:13.198000Z'
}

{
  processing_time: 91.141,
  results: [
    {
      box: { xmin: 292, ymin: 372, xmax: 496, ymax: 444 },
      plate: 'nsn666',
      region: { score: 0.94, code: 'de' },
      vehicle: {
        score: 0.808,
        box: { xmin: 7, ymin: 79, xmax: 652, ymax: 561 },
        type: 'Car'
      },
      score: 0.901,
      candidates: [ { score: 0.901, plate: 'nsn666' } ],
      dscore: 0.923
    },
    {
      box: { xmin: 656, ymin: 206, xmax: 720, ymax: 228 },
      plate: 'nsn458',
      region: { score: 0.882, code: 'de' },
      vehicle: {
        score: 0.358,
        box: { xmin: 430, ymin: 90, xmax: 788, ymax: 279 },
        type: 'Car'
      },
      score: 0.904,
      candidates: [
        { score: 0.904, plate: 'nsn458' },
        { score: 0.763, plate: 'nsn45b' }
      ],
      dscore: 0.728
    },
    {
      box: { xmin: 818, ymin: 196, xmax: 851, ymax: 223 },
      plate: 'ak92115',
      region: { score: 0.852, code: 'it' },
      vehicle: {
        score: 0,
        type: 'Unknown',
        box: { xmin: 0, ymin: 0, xmax: 0, ymax: 0 }
      },
      score: 0.91,
      candidates: [
        { score: 0.91, plate: 'ak92115' },
        { score: 0.789, plate: 'ak921i5' },
        { score: 0.784, plate: 'ak92i15' },
        { score: 0.663, plate: 'ak92ii5' }
      ],
      dscore: 0.428
    }
  ],
  filename: '0842_6gW1A_b13d0a9b-b774-47c4-b80d-cc88b0d0e317.jpg',
  version: 1,
  camera_id: 'homey',
  timestamp: '2020-05-04T08:42:18.134000Z'
}

*/
